calc/l3_route_resolver.go

```diff
@@ -151,38 +178,91 @@ func (c *L3RouteResolver) OnBlockUpdate(update api.Update) (_ bool) {
 }
 
 func (c *L3RouteResolver) OnResourceUpdate(update api.Update) (_ bool) {
+	// We only care about nodes, not other resources.
 	resourceKey := update.Key.(model.ResourceKey)
 	if resourceKey.Kind != apiv3.KindNode {
 		return
 	}
 
+	// Queue up a flush.
+	defer c.flush()
+
+	// Extract the nodename and check whether the node was known already.
 	nodeName := update.Key.(model.ResourceKey).Name
+	_, nodeExisted := c.nodeNameToNode[nodeName]
+
 	logCxt := logrus.WithField("node", nodeName).WithField("update", update)
 	logCxt.Debug("OnResourceUpdate triggered")
+	var myOldCIDR, myNewCIDR *cnet.IPNet
+	if nodeName == c.myNodeName {
+		// Our node, look up our old CIDR to see if it changes.
+		logCxt.Debug("Update to this node's resource.")
+		myOldCIDR, _ = c.nodeCidr(c.myNodeName)
+	}
+
+	// Update our tracking data structures.
 	if update.Value != nil && update.Value.(*apiv3.Node).Spec.BGP != nil {
 		node := update.Value.(*apiv3.Node)
 		bgp := node.Spec.BGP
 		c.nodeNameToNode[nodeName] = node
 		ipv4, _, err := cnet.ParseCIDROrIP(bgp.IPv4Address)
 		if err != nil {
+			// Validation should prevent this...
 			logCxt.WithError(err).Error("couldn't parse ipv4 address from node bgp info")
-			return
+			if nodeExisted {
+				logCxt.WithError(err).Error("Treating as deletion")
+				delete(c.nodeNameToNode, nodeName)
+				c.onRemoveNode(nodeName)
+			}
```

Comment:
I was just copying the existing behaviour.  Thinking about it; I think we can have a BGP spec that is non-nil and has an empty IP, which will fall through this case.  I'll adjust the logic so that we only try to parse if it's not "" and then upgrade to a panic.